torch_geometric/transforms/mesh.py

```diff
@@ -0,0 +1,311 @@
+from tempfile import mkstemp
+from shutil import move
+import torch
+import numpy as np
+import os
+from ..nn.mesh.mesh_union import MeshUnion
+import torch_geometric.data
+
+
+class MeshData(torch_geometric.data.Data):
+    r"""MeshData class which holds a mesh data for implementing `"MeshCNN: A
+    Network with an Edge" <https://arxiv.org/abs/1809.05910>`_ paper
+    This class is a torch geometric Data extension.
+
+    Attributes:
+        filename (str): obj file name
+        vs (np.array(Num_vertices, 3, dtype=float64)): vertices positions
+        edges (np.array(Num_edges, 2, dtype=int32)): pairs of vertices indexes
+                                                     which construct an edge
+        sides (np.array(Num_edges, 4, dtype=int32): values of: 0,1,2,3 -
+                                                    sides[E][N] indicates the
+                                                    index of edge E for the N
+                                                    neighbour.
+        edges_count (int): number of edges in the mesh obj.
+        ve (np.array(Num_vertices,
+        Num_connections, dtype=int32)): array of all edge id connections of
+                                        vertex i e.g., ve[i,:] = [edge_ind_j,
+                                        edge_ind_k] means vertex i have 2
+                                        connections with vertices according to
+                                        edges[edge_ind_j, :] and
+                                        edges[edge_ind_k, :]
+        v_mask (np.array(Num_vetices, dtype=bool): indicates if a vertex
+                                                   still in the mesh obj
+                                                   (True - exists,
+                                                   False - removed)
+        edge_lengths (np.array(Num_edges, dtype=float64): edges L2 norm
+        edge_areas (np.array(Num_edges,
+                   dtype=float64): for each edge it is the sum of 1/3
+                                   face(s) area the edge constructs
+                                   (maximum 2 faces), normalized by the total
+                                   area of all the faces. Used for
+                                   segmentation accuracy.
+        edge_index (np.array(2, Num_edges x 5,
```

Comment:
How about we hold the mesh graph data in `edge_index` (just as in standard PyG), and have an additional index matrix `neighbor_e_id` of shape `[num_edges, 4]` that holds the indices of direct neighbors for every edge?

Reply:
I am not sure I understand. I was followed by the introduction [example](https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html):

edge_index = torch.tensor([[0, 1, 1, 2], [1, 0, 2, 1]], dtype=torch.long)

Can you please elaborate?

Reply:
Maybe I misunderstood the code, but my impression is that you have a regular `edge_index` tensor of shape `[2, num_edges]` denoting mesh connections. For MeshCNN, we need to collect the direct neighboring edges, *e.g.*, 4 edges in case of two neighboring faces. I think one can just create a tensor of shape `[num_edges, 4]` that hold the indices of edges denoted in `edge_index` that are attached to the specific edge. WDYT?